health-services/ExerciseSampleCompose/app/src/main/java/com/example/exercisesamplecompose/service/ExerciseService.kt

```diff
@@ -0,0 +1,422 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.example.exercisesamplecompose.service
+
+import android.app.Notification
+import android.app.NotificationChannel
+import android.app.NotificationManager
+import android.app.PendingIntent
+import android.app.Service
+import android.content.ContentValues.TAG
+import android.content.Intent
+import android.os.Binder
+import android.os.IBinder
+import android.os.SystemClock
+import android.util.Log
+import androidx.core.app.NotificationCompat
+import androidx.health.services.client.data.DataPointContainer
+import androidx.health.services.client.data.ExerciseEndReason
+import androidx.health.services.client.data.ExerciseState
+import androidx.health.services.client.data.ExerciseUpdate
+import androidx.health.services.client.data.LocationAvailability
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleService
+import androidx.lifecycle.lifecycleScope
+import androidx.lifecycle.repeatOnLifecycle
+import androidx.wear.ongoing.OngoingActivity
+import androidx.wear.ongoing.Status
+import com.example.exercisesamplecompose.MainActivity
+import com.example.exercisesamplecompose.R
+import com.example.exercisesamplecompose.data.ExerciseClientManager
+import com.example.exercisesamplecompose.data.ExerciseMessage
+import com.example.exercisesamplecompose.data.HealthServicesRepository
+import dagger.hilt.android.AndroidEntryPoint
+import java.time.Duration
+import java.time.Instant
+import javax.inject.Inject
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+
+
+@AndroidEntryPoint
+class ForegroundService : LifecycleService() {
+
+    @Inject lateinit var exerciseClientManager: ExerciseClientManager
+    @Inject lateinit var healthServicesRepository: HealthServicesRepository
+
+    private var isBound = false
+    private var isStarted = false
+    private val localBinder = LocalBinder()
+    private var serviceRunningInForeground = false
+
+    private val _exerciseState = MutableStateFlow(ExerciseState.ENDED)
+    val exerciseState: StateFlow<ExerciseState> = _exerciseState
+
+    private val _exerciseMetrics = MutableStateFlow<DataPointContainer?>(null)
+    val exerciseMetrics: StateFlow<DataPointContainer?> = _exerciseMetrics
+
+    private val _exerciseLaps = MutableStateFlow(0)
+    val exerciseLaps: StateFlow<Int> = _exerciseLaps
+
+    private val _exerciseDurationUpdate = MutableStateFlow<ActiveDurationUpdate?>(null)
+    val exerciseDurationUpdate: StateFlow<ActiveDurationUpdate?> = _exerciseDurationUpdate
+
+    private val _exerciseStateChange = MutableStateFlow<ExerciseStateChange>(
+        ExerciseStateChange.OtherStateChange(
+            ExerciseState.ENDED
+        )
+    )
+    val exerciseStateChange: StateFlow<ExerciseStateChange> = _exerciseStateChange
```

Comment:
Since consumers of this service only see the `ExcerciseStateChange`, we can keep all the mutable values in here in a single `MutableStateFlow`.


```kotlin
data class ExerciseServiceState(
    val exerciseState: ExerciseState> = ExerciseState.ENDED,
    val exerciseMetrics: DataPointContainer? = null,
    val exerciseLaps: Int = 0,
    val exerciseDurationUpdate: ActiveDurationUpdate? = null,
)
```

Then

```kotlin
private val exserciseServiceState: MutableStateFlow(ExerciseServiceState())
```

Then to update each variable, you can use

```kotlin
excerciseServiceState.update { oldState ->
    olState.copy(...) // update as many fields as you want to here
}
```

This way you don't have to wrangle so many fields.

Reply:
Thanks TJ! Have flattened this